File: src/org/cerion/symcalc/expression/function/integer/OddQ.java
package org.cerion.symcalc.expression.function.integer;

import org.cerion.symcalc.exception.ValidationException;
import org.cerion.symcalc.expression.BoolExpr;
import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.number.IntegerNum;

public class OddQ extends FunctionExpr {

    public OddQ(Expr ...e) {
        super(FunctionType.ODDQ, e);
    }

    @Override
    protected Expr evaluate() {

        if(!get(0).isInteger())
            return BoolExpr.FALSE;

        IntegerNum N = (IntegerNum)get(0);
        if(N.IsOdd())
            return BoolExpr.TRUE;

        return BoolExpr.FALSE;
    }

    @Override
    public void validate() throws ValidationException {
        validateParameterCount(1);
    }
}

File: src/org/cerion/symcalc/expression/function/integer/OddQ.kt
package org.cerion.symcalc.expression.function.integer

import org.cerion.symcalc.exception.ValidationException
import org.cerion.symcalc.expression.BoolExpr
import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.function.Function
import org.cerion.symcalc.expression.function.FunctionExpr

class OddQ(vararg e: Expr) : FunctionExpr(Function.ODDQ, *e) {

    override fun evaluate(): Expr {
        if (!get(0).isInteger)
            return BoolExpr.FALSE

        return if (get(0).asInteger().isOdd) BoolExpr.TRUE else BoolExpr.FALSE
    }

    @Throws(ValidationException::class)
    override fun validate() {
        validateParameterCount(1)
    }
}
